apm-agent-core/src/main/java/co/elastic/apm/agent/bci/ElasticApmAgent.java

```diff
@@ -348,7 +396,7 @@ private static AgentBuilder getAgentBuilder(final ByteBuddy byteBuddy, final Cor
 
         return new AgentBuilder.Default(byteBuddy)
             .with(AgentBuilder.RedefinitionStrategy.RETRANSFORMATION)
```

Comment:
```suggestion
            .with(AgentBuilder.RedefinitionStrategy.RETRANSFORMATION)
            .with(new AgentBuilder.RedefinitionStrategy.BatchAllocator() {
                @Override
                public Iterable<? extends List<Class<?>>> batch(List<Class<?>> types) {
                    // if this is on startup, do the same as BatchAllocator.ForTotal, otherwise use batches
                }
            })
```
Not exactly this, but this `BatchAllocator` can be handy for retransformations...

Reply:
Byte Buddy is batching by default (via `RedefinitionStrategy.BatchAllocator.ForTotal.INSTANCE`)

Reply:
`RedefinitionStrategy.BatchAllocator.ForTotal.INSTANCE` means it is not batching, it includes all in one batch

Reply:
I see what you mean now. Which batch size would you propose? Should we sleep between batches? Or should we just start with a single batch?

Reply:
I would say it makes sense to use a single batch on startup when the instrumentations are applied through `premain` or programmatic attach and use batches when applied through `agentmain` or when re-instrumenting. 

> should we just start with a single batch?

The reason I am suggesting it in this PR is because of the introduction of re-instrumentation, but if you have a reason to merge this quickly, it can be done in a separate PR. I think that if it is a matter of small effort- it worth it. Your call.

> Which batch size would you propose? 

Well, as you imagine, it would just be random anyways 🙂 . The main idea is not to cause a long degradation in app performance. From the top of my head- 100 per batch.

> Should we sleep between batches?

Certainly, . We can use `net.bytebuddy.agent.builder.AgentBuilder.RedefinitionStrategy.Listener.Pausing` for that.